src/assignment-cache.ts

```diff
@@ -0,0 +1,76 @@
+import { LRUCache } from 'lru-cache';
+
+import { EppoValue } from './eppo_value';
+
+export interface AssignmentCacheKey {
+  subjectKey: string;
+  flagKey: string;
+  allocationKey: string;
+  variationValue: EppoValue;
+}
+
+interface Cacheable {
+  get(key: string): string | undefined;
+  set(key: string, value: string): void;
+  has(key: string): boolean;
+}
+
+export abstract class AssignmentCache<T extends Cacheable> {
+  // key -> variation value hash
+  protected cache: T;
+
+  constructor(cacheInstance: T) {
+    this.cache = cacheInstance;
+  }
+
+  hasLoggedAssignment(key: AssignmentCacheKey): boolean {
+    // no cache key present
+    if (!this.cache.has(this.getCacheKey(key))) {
+      return false;
+    }
+
+    // the subject has been assigned to a different variation
+    // than was previously logged.
+    // in this case we need to log the assignment again.
+    if (this.cache.get(this.getCacheKey(key)) !== key.variationValue.toHashedString()) {
+      return false;
+    }
```

Comment:
@petzel I am debating whether it is correct to also remove the cache entry inside this `if` statement.

Right now it feels like the cache "wrong" and we are relying on the `logAssignment` method being called later to over-write the cache value to the newly seen variation value, even know we know about it already.

Feedback appreciated.